wsbroadcastserver/clientconnection.go

```diff
@@ -76,49 +101,167 @@ func (cc *ClientConnection) Compression() bool {
+	}
+	return nil
+}
+
+func (cc *ClientConnection) writeBroadcastMessage(bm *m.BroadcastMessage) error {
+	notCompressed, compressed, err := serializeMessage(bm, !cc.compression, cc.compression)
+	if err != nil {
+		return err
+	}
+
+	var data []byte
+	if cc.compression {
+		data = compressed.Bytes()
+	} else {
+		data = notCompressed.Bytes()
+	}
+	err = cc.writeRaw(data)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
 func (cc *ClientConnection) Start(parentCtx context.Context) {
 	cc.StopWaiter.Start(parentCtx, cc)
 	cc.LaunchThread(func(ctx context.Context) {
+		// A delay may be configured, ensures the Broadcaster delays before any
+		// messages are sent to the client. The ClientConnection has not been
+		// registered so the out channel filling is not a concern.
 		if cc.delay != 0 {
-			var delayQueue [][]byte
 			t := time.NewTimer(cc.delay)
-			done := false
-			for !done {
-				select {
-				case <-ctx.Done():
-					return
-				case data := <-cc.out:
-					delayQueue = append(delayQueue, data)
-				case <-t.C:
-					for _, data := range delayQueue {
-						err := cc.writeRaw(data)
-						if err != nil {
-							logWarn(err, "error writing data to client")
-							cc.clientManager.Remove(cc)
-							return
-						}
-					}
-					done = true
-				}
+			select {
+			case <-ctx.Done():
+				return
+			case <-t.C:
 			}
 		}
 
+		// Send the current backlog before registering the ClientConnection in
+		// case the backlog is very large
+		segment := cc.backlog.Head()
+		if !backlog.IsBacklogSegmentNil(segment) && segment.Start() < uint64(cc.requestedSeqNum) {
+			s, err := cc.backlog.Lookup(uint64(cc.requestedSeqNum))
+			if err != nil {
+				logWarn(err, "error finding requested sequence number in backlog: sending the entire backlog instead")
+			} else {
+				segment = s
+			}
+		}
+		err := cc.writeBacklog(ctx, segment)
+		if errors.Is(err, errContextDone) {
+			return
+		} else if err != nil {
+			logWarn(err, "error writing messages from backlog")
+			cc.Remove()
+			return
+		}
+
+		cc.Register()
+		timer := time.NewTimer(5 * time.Second)
+		select {
+		case <-ctx.Done():
+			return
+		case <-cc.registered:
+			log.Debug("ClientConnection registered with ClientManager", "client", cc.Name)
+		case <-timer.C:
+			log.Error("timed out waiting for ClientConnection to register with ClientManager", "client", cc.Name)
+		}
+
+		// broadcast any new messages sent to the out channel
 		for {
 			select {
 			case <-ctx.Done():
 				return
-			case data := <-cc.out:
-				err := cc.writeRaw(data)
+			case msg := <-cc.out:
+				expSeqNum := cc.LastSentSeqNum.Load() + 1
+				if !cc.backlogSent && msg.sequenceNumber != nil && uint64(*msg.sequenceNumber) > expSeqNum {
+					catchupSeqNum := uint64(*msg.sequenceNumber) - 1
+					bm, err := cc.backlog.Get(expSeqNum, catchupSeqNum)
+					if err != nil {
+						logWarn(err, fmt.Sprintf("error reading messages %d to %d from backlog", expSeqNum, catchupSeqNum))
+						cc.Remove()
+						return
```

Comment:
I don't think this is the client's fault. At this point I think we should just log the warning but not disconnect the client. That should mean that they'll eventually get in sync with the feed at least.